docs/proposals/odo-debug.md

```diff
@@ -0,0 +1,82 @@
+# odo debug support proposal
+
+Table of Contents:
+- [odo debug support proposal](#odo-debug-support-proposal)
+  - [Design overview](#design-overview)
+    - [New odo commands](#new-odo-commands)
+      - [`odo debug start`](#odo-debug-start)
+      - [`odo debug stop`](#odo-debug-stop)
+      - [`odo debug status`](#odo-debug-status)
+    - [`odo debug port-forward LOCAL_PORT:REMOTE_PORT`](#odo-debug-port-forward-localportremoteport)
+- [Languages support](#languages-support)
+  - [NodeJS](#nodejs)
+    - [Problem 1](#problem-1)
+      - [Solution 1](#solution-1)
+      - [Solution 2](#solution-2)
+  - [Java](#java)
+
+## Design overview
+
+The debug mode inside the container is controlled by `DEBUG_MODE` and `DEBUG_PORT` env variable.
+- if `DEBUG_MODE == true` than the code should be executed with remote debugging capabilities enabled.
+- `DEBUG_PORT` is optional and controls the port of the remote debugger.
+
+Odo sets up port-forwarding to allow a debugger to connect to the running process. 
+Debugger client will always connect to `localhost:$DEBUG_PORT`
+### New odo commands
+#### `odo debug start`
+Enables debug mode for the component. It does this by adding `DEBUG_MODE = true` env variable.
+
+After adding env variables the component (Deployment Config) is restarted.
+Once the pod is in running state `odo debug start` executes `odo debug port-forward $DEBUG_PORT:$DEBUG_PORT`.
+This process executed on the background and PID of the process is recorded in `~/.odo/port-forward.pid` file.
```

Comment:
> Could the creation of `~/.odo/port-forward.pid` fail under any circumstances?
> * If yes, what would `odo` do?

fail



> Is this information really going to be stored in `~/.odo/port-forward.pid` file (global configuration directory) or in a component directory `/path/to/component/.odo/port-forward.pid`?
> * If it's going to be in `~/.odo/port-forward.pid` indeed, wouldn't that limit the developer to debug only one app at a time? What if they want to start another component in debug mode?

good question. Yes, it would limit it to debugging just one application at a time.
Problem with storing it in components local .odo dir is that users will tend to commit this into the source code :-( And this is not something that should be persistent, it is runtime information.

Maybe storing pids in the global configuration in json format?
```json
[
  {
     "directory": "/path/to/component/",
     "port-forward-pid": 344
  },
  {
     "directory": "/path/to/another/component/",
     "port-forward-pid": 332
  }
]
```

Reply:
Makes sense to me. Thanks for the clarification. :+1:

Reply:
For the first question: so we intend to have the operation of enabling port-forwarding and creation/modification of the file as an atomic action, right?

Reply:
Yes. from the outside ti should look like atomic operation.
If any of the file operations fails, the `port-forward` process needs to be killed, and error reported.


Implementation `odo debug start` could be something like this:

- check if `port-forward-pids.json` has a pid record for the component
- if yes and the process with given PID exists, stop and show a message that the port forwarding is already running
- if yes and the process is not running, show a warning that the process might fail and this will execute the new one
- set environment variables
- wait for deployment
- execute `odo debug port-forward`, get its pid
- record pid to the port-forward-pids.json (create file if doesn't exist)
- if any of the file operations fails, kill `odo debug port-forward` process and error out.

Reply:
@kadel this sound awesome to me! Thanks for the detailed write-up. It's going to be pretty easy for whoever picks up this piece to code.